### PR TITLE
[Chore] Add `withRoles` method to user factory

### LINKED ISSUE #6407: ♻️ Add "states" to UserFactory for roles
## ♻️ Debt/Refactor

Add [state functions](https://laravel.com/docs/10.x/eloquent-factories#factory-states) to UserFactory to simplify creating users with particular roles. 

## 🕵️ Details

We often have to create a user with particular roles in seeders and in tests, and it would be nice to have a consistent, easy way of doing so.

## 🙋‍♀️ Proposed Solution

(optional) Do you have a proposed solution?

## 🌎 Localization

(optional) Provide any new copy along with translations available.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the debt was addressed and expected functionality has not been affected.

- [ ] UserFactory has shortcut 'state' functions for creating users with specific roles
- [ ] PhpUnit tests and seeders use the state functions consistently to create users with specific roles

### PR DESCRIPTION
**Pull request by esizer**
🤖 Resolves #6407 

## 👋 Introduction

This adds a `withRoles` method to the user factory to attach roles after creation.

### Usage

```php
// Attach roles with no team
User::factory()->withRoles(["base_user", "applicant"]);

// Attach roles to a team
User::factory()->withRoles(["pool_operator"], "team_name");

// Attach roles to multiple teams
User::factory()->withRoles(["pool_operator"], ["team_name", "second_team_name"]);

// Chain method (for non-team and team roles
User::factory()
  ->withRoles(["base_user", "applicant"])
  ->withRoles(["pool_operator"], "team_name");
```

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Run a fresh seed `docker-compose exec -w /home/site/wwwroot/api webserver sh -c "php artisan migrate:fresh --seed"`
2. Confirm all users have to appropriate roles
3. Run PHP tests `docker-compose exec -w /home/site/wwwroot/api webserver sh -c "php artisan test"`
4. Confirm all tests pass (excluding the known issues)

### CODE REVIEW

File: api/database/factories/UserFactory.php [hunk 1]
```diff
@@ -209,4 +209,29 @@ public function configure()
...
+     * Attach roles to a user after creation.
+     *
+     * @param array $roles Array of the roles to attach
+     * @param string $team Name of the team to attach the roles to
+     *
+     * @return $this
+     *
+     * Usage:
+     *      `User::factory()->withRoles(['base_user', 'applicant']);`
+     *      `User::factory()->withRoles(['poo_operator'], 'team_name');`
```
Comment (petertgiles) [inline api/database/factories/UserFactory.php:223]: lol :hankey:

**Review by petertgiles — APPROVED** [review]
Lol, it's pretty much the same function to use though I guess you've hit acceptance criteria.  It is nice to be able to hit more than one team in a single statement I guess.  It might be worth waiting until @tristan-orourke gets back to make sure we're not misinterpreting the isssue.

File: api/database/factories/UserFactory.php (hunk 1 above)
Comment (esizer) [reply api/database/factories/UserFactory.php]: oooopsss 😅 Fixed in [fix typo](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6663/commits/3a0e583a83aa8cf9b217a338f4df25eca3a34325)

**Comment by esizer** [comment]
> Lol, it's pretty much the same function to use though I guess you've hit acceptance criteria. It is nice to be able to hit more than one team in a single statement I guess. It might be worth waiting until @tristan-orourke gets back to make sure we're not misinterpreting the isssue.

I was kind of thinking the same thing as I did this but I couldn't really think of any other way to do it 🤔 The only other way I could think of would be just as verbose (maybe more?)

```php
//Current
User::factory()
  ->withRoles(['base_user', 'applicant'])
  ->withRoles(['pool_operator'], 'team_name');

// Versus
User::factory()
  ->asBaseUser()
  ->asApplicant()
  ->asAdmin()
  ->asPoolOperator('team_name')
```

**Comment by tristan-orourke** [comment]
Hah, I see what you mean, it really isn't much less code...

Personally, I really like your suggestion here:
```
User::factory()
  ->asBaseUser()
  ->asApplicant()
  ->asAdmin()
  ->asPoolOperator('team_name')
```
That makes the team arg explicitely required, which I like, while still matching the other roles. 

The other thing I would do here is add the base_user role automatically when creating a user (can be done with [afterCreating](https://laravel.com/docs/10.x/eloquent-factories#factory-callbacks) in the factory) instead of adding it via the state.

**Comment by esizer** [comment]
> Hah, I see what you mean, it really isn't much less code...
> 
> Personally, I really like your suggestion here:
> 
> ```
> User::factory()
>   ->asBaseUser()
>   ->asApplicant()
>   ->asAdmin()
>   ->asPoolOperator('team_name')
> ```
> 
> That makes the team arg explicitely required, which I like, while still matching the other roles.
> 
> The other thing I would do here is add the base_user role automatically when creating a user (can be done with [afterCreating](https://laravel.com/docs/10.x/eloquent-factories#factory-callbacks) in the factory) instead of adding it via the state.

Okay, updated with the new implementation and fixed any tests that broke with the new default role.

- [update tests and factory with new role methods](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6663/commits/7473319d9e2804960de3afef5afe51e24d0ea61a)
- [fix remaining tests with new base role](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6663/commits/60bbc2887f12404e46b14e461a73bbd600caaa66)

**Review by tristan-orourke — APPROVED** [review]

**Pull request merged**